internal/querynodev2/segments/result.go

```diff
@@ -524,25 +538,80 @@ func MergeSegcoreRetrieveResults(ctx context.Context, retrieveResults []*segcore
 		pk := typeutil.GetPK(validRetrieveResults[sel].GetIds(), cursors[sel])
 		if _, ok := idSet[pk]; !ok {
 			typeutil.AppendPKs(ret.Ids, pk)
-			retSize += typeutil.AppendFieldData(ret.FieldsData, validRetrieveResults[sel].GetFieldsData(), cursors[sel])
+			selected = append(selected, sel)
+			selectedOffsets[sel] = append(selectedOffsets[sel], validRetrieveResults[sel].GetOffset()[cursors[sel]])
+			selectedIndexes[sel] = append(selectedIndexes[sel], cursors[sel])
 			idSet[pk] = struct{}{}
 		} else {
 			// primary keys duplicate
 			skipDupCnt++
 		}
 
-		// limit retrieve result to avoid oom
-		if retSize > maxOutputSize {
-			return nil, fmt.Errorf("query results exceed the maxOutputSize Limit %d", maxOutputSize)
-		}
-
 		cursors[sel]++
 	}
 
 	if skipDupCnt > 0 {
 		log.Debug("skip duplicated query result while reducing segcore.RetrieveResults", zap.Int64("dupCount", skipDupCnt))
 	}
 
+	if !plan.ignoreNonPk {
+		// target entry already retrieved, don't do this after AppendPKs for better performance. Save the cost everytime
+		// judge the `!plan.ignoreNonPk` condition.
+		_, span2 := otel.Tracer(typeutil.QueryNodeRole).Start(ctx, "MergeSegcoreResults-AppendFieldData")
+		defer span2.End()
+		ret.FieldsData = make([]*schemapb.FieldData, len(validRetrieveResults[0].GetFieldsData()))
+		cursors = make([]int64, len(validRetrieveResults))
+		for _, sel := range selected {
+			// cannot use `cursors[sel]` directly, since some of them may be skipped.
+			retSize += typeutil.AppendFieldData(ret.FieldsData, validRetrieveResults[sel].GetFieldsData(), selectedIndexes[sel][cursors[sel]])
+			cursors[sel]++
+		}
+	} else {
+		// target entry not retrieved.
+		ctx, span2 := otel.Tracer(typeutil.QueryNodeRole).Start(ctx, "MergeSegcoreResults-RetrieveByOffsets-AppendFieldData")
+		defer span2.End()
+		segmentResults := make([]*segcorepb.RetrieveResults, len(validRetrieveResults))
+		eg, ctx := errgroup.WithContext(ctx)
+		for i, offsets := range selectedOffsets {
+			if len(offsets) == 0 {
+				log.Ctx(ctx).Debug("skip empty retrieve results", zap.Int64("segment", validSegments[i].ID()))
+				continue
+			}
+			idx, theOffsets := i, offsets
+			eg.Go(func() error {
```

Comment:
use thread pool instead

Reply:
Should we use which pool to schedule this?

Reply:
Btw, the inner RetrieveByOffsets was already scheduled by the query thread pool.

Reply:
I don't find we take any concurrent control on `retrieveOnSegments`, neither.